ndonnx/_typed_array/funcs.py

```diff
@@ -0,0 +1,313 @@
+# Copyright (c) QuantCo 2023-2025
+# SPDX-License-Identifier: BSD-3-Clause
+from __future__ import annotations
+
+from collections.abc import Sequence
+from functools import reduce
+from types import NotImplementedType
+from typing import TYPE_CHECKING, Literal, TypeVar, overload
+
+import numpy as np
+from spox import Var
+
+from .._dtypes import DType
+from .._types import NestedSequence, PyScalar
+from .typed_array import TyArrayBase
+from .utils import promote
+
+if TYPE_CHECKING:
+    from .. import Array
+    from .._dtypes import TY_ARRAY_BASE
+    from . import onnx
+
+
+def _infer_sequence(
+    val: NestedSequence,
+) -> DType:
+    from . import onnx
+
+    types = set()
+    for item in val:
+        if isinstance(item, Sequence) and not isinstance(item, str):
+            types.add(_infer_sequence(item))
+        else:
+            types.add(_infer_dtype(item))
+    if len(types) == 0:
+        return onnx.float64
+    return _result_dtype(*types)
+
+
+def _infer_dtype(
+    val: PyScalar | np.ndarray | TyArrayBase | Var | NestedSequence,
+) -> DType:
+    from . import masked_onnx, onnx
+
+    if isinstance(val, np.ndarray):
+        if val.dtype == object:
+            if not all(isinstance(el, str) for el in val.flatten()):
+                raise ValueError(
+                    "found array with object data type but it contains non-string elements"
+                )
+            return onnx.utf8
+        core_type = onnx.from_numpy(val.dtype)
+        if isinstance(val, np.ma.MaskedArray):
+            return masked_onnx.to_nullable_dtype(core_type)
+        else:
+            return core_type
+    elif isinstance(val, TyArrayBase):
+        return val.dtype
+    elif isinstance(val, Var):
+        return onnx.from_numpy(val.unwrap_tensor().dtype)
+    elif isinstance(val, bool):
+        return onnx.bool_
+    elif isinstance(val, int):
+        return onnx.int64
+    elif isinstance(val, float):
+        return onnx.float64
+    elif isinstance(val, str):
+        return onnx.utf8
+    elif isinstance(val, Sequence):
+        return _infer_sequence(val)
+    elif isinstance(val, np.generic):
+        return onnx.from_numpy(val.dtype)
+    else:
+        raise ValueError(f"Unable to infer dtype from {val}")
+
+
+@overload
+def astyarray(
+    val: PyScalar | np.ndarray | TyArrayBase | Var | Array | NestedSequence,
+    dtype: DType[TY_ARRAY_BASE],
+) -> TY_ARRAY_BASE: ...
+
+
+@overload
+def astyarray(
+    val: PyScalar | np.ndarray | TyArrayBase | Var | Array | NestedSequence,
+    dtype: None | DType = None,
+) -> TyArrayBase: ...
+
+
+def astyarray(
+    val: PyScalar | np.ndarray | TyArrayBase | Var | Array | NestedSequence,
+    dtype: None | DType[TY_ARRAY_BASE] = None,
+) -> TyArrayBase:
```

Comment:
We should aim to have acyclic dependencies. This can be achieved by having each "abstraction" level only dependent on items at the same level or lower-level. Here, this function shouldn't accept an `Array` but only a `TyArrayBase` and the lazy import goes away. The caller can handle unwrapping the `Array`.